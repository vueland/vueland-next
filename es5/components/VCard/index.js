"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "VCard", {
  enumerable: true,
  get: function get() {
    return _VCard.VCard;
  }
});
exports.VCardActions = exports.VCardContent = exports.VCardSubtitle = exports.VCardTitle = void 0;

var _helpers = require("../../helpers");

var _VCard = require("./VCard");

var VCardTitle = (0, _helpers.createSimpleFunctional)("v-card__title");
exports.VCardTitle = VCardTitle;
var VCardSubtitle = (0, _helpers.createSimpleFunctional)("v-card__subtitle");
exports.VCardSubtitle = VCardSubtitle;
var VCardContent = (0, _helpers.createSimpleFunctional)("v-card__content");
exports.VCardContent = VCardContent;
var VCardActions = (0, _helpers.createSimpleFunctional)("v-card__actions");
exports.VCardActions = VCardActions;
//# sourceMappingURL=index.js.map