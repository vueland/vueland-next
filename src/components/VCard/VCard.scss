@import "../../styles/scss/modules/global";

.v-card {
  @include flex(center, center, column);
  width: auto;
  background: $mainWhite;
  border-radius: 6px;
  overflow: hidden;
  position: relative;
  z-index: 3;

  &__title {
    @include flex(flex, flex-start, center);
    width: 100%;
    padding: 5px 20px;
    min-height: 45px;
    font-size: 1.4rem;
    position: relative;
    font-family: $mainFont;
  }

  &__content {
    @include flex(flex, center, center, column);
    text-align: left;
    width: 100%;
    min-height: 100px;
    font-size: 1rem;
    font-family: $mainFont;
    padding: 20px;
    line-height: normal;
  }

  &__actions {
    @include flex(flex, space-evenly, center);
    width: 100%;
    min-height: 60px;
    padding: 0 20px;
    position: relative;
  }
}
