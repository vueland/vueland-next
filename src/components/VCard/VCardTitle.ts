import { createSimpleFunctional } from '../../helpers'

export default createSimpleFunctional('v-card__title')
