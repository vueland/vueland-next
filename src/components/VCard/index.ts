import { createSimpleFunctional } from "../../helpers";

export { VCard } from "./VCard";

export const VCardTitle = createSimpleFunctional("v-card__title");
export const VCardSubtitle = createSimpleFunctional("v-card__subtitle");
export const VCardContent = createSimpleFunctional("v-card__content");
export const VCardActions = createSimpleFunctional("v-card__actions");
