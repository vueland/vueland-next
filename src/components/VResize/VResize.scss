.v-resize {
  display: flex;
  position: absolute;
  background: rgba(239, 236, 236, .2);
  z-index: 10;

  &--top {
    top: 0;
    left: 0;
    width: 100%;
    height: 5px;
    cursor: row-resize;
  }

  &--bottom {
    bottom: 0;
    left: 0;
    width: 100%;
    height: 5px;
    cursor: row-resize;
  }

  &--right {
    top: 0;
    right: 0;
    width: 5px;
    height: 100%;
    cursor: col-resize;
  }

  &--left {
    top: 0;
    left: 0;
    width: 5px;
    height: 100%;
    cursor: col-resize;
  }
}
