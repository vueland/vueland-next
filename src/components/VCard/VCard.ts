// Styles
import './VCard.scss'

// Vue API
import { defineComponent, h, computed } from 'vue'

// Compositions
import { colorProps, useColors } from '../../effects/use-colors'
import { elevationProps, useElevation } from '../../effects/use-elevation'

// Types
import { Props } from '../../types'
import { VNode } from 'vue'

const cardProps: Props = {
  width: [String, Number],
  ...colorProps(),
  ...elevationProps(),
}

export const VCard = defineComponent({
  props: cardProps,

  setup(props, { slots }): () => VNode {
    const { setBackground } = useColors()
    const { elevationClasses } = useElevation(props)

    const classes = computed(
      (): Record<string, boolean> => {
        return {
          'v-card': true,
          ...elevationClasses.value,
        }
      },
    )

    const dataProps = () => {
      return {
        class: classes.value,
        style: { maxWidth: `${props.width}px` },
      }
    }

    return () =>
      h(
        'div',
        props.color ? setBackground(props.color, dataProps()) : dataProps(),
        slots.default && slots.default(),
      )
  },
})
