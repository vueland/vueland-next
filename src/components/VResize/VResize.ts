// Styles
import './VResize.scss'

// Vue API
import {
  defineComponent,
  h,
  reactive,
  ref,
  computed,
  onMounted,
} from 'vue'

// Effects
import { positionProps } from '../../effects/use-position'

// Types
import { Props } from '../../types'

const vResizeProps: Props = {
  emit: {
    type: Boolean,
    default: false,
  },

  customClass: {
    type: String,
  },

  minSize: {
    type: [String, Number],
    default: 50,
  },
  ...positionProps(),
}

type ResizeData = {
  parentNode: HTMLElement | null
  offsetTop: number
  offsetLeft: number
  parentHeight: number
  parentWidth: number
  marginLeft: number
  marginTop: number
  resized: boolean
}

export const VResize = defineComponent({
  name: 'v-resize',
  props: vResizeProps,

  setup(props, { emit }) {

    const data: ResizeData = reactive({
      parentNode: null,
      offsetTop: 0,
      offsetLeft: 0,
      parentHeight: 0,
      parentWidth: 0,
      marginLeft: 0,
      marginTop: 0,
      resized: false,
    })

    const resRef = ref<HTMLElement | null>(null)

    const classes = computed<Record<string, boolean>>(() => {
      return {
        'v-resize': true,
        'v-resize--top': props.top,
        'v-resize--bottom': props.bottom,
        'v-resize--right': props.right,
        'v-resize--left': props.left,
        [props.customClass]: !!props.customClass,
      }
    })

    const isDirectY = computed<boolean>(() => {
      return props.top || props.bottom
    })

    const isNeedReverse = computed<boolean>(() => {
      return props.top || props.left
    })

    const currentSize = computed<number>(() => {
      return isDirectY.value ? data.parentHeight! : data.parentWidth!
    })

    const styleProp = computed<string>(() => {
      return isDirectY.value ? 'height' : 'width'
    })

    const reverseFrom = computed<string>(() => {
      return props.top ? 'top' : 'left'
    })

    const offset = computed<number>(() => {
      return isDirectY.value ? data.offsetTop! : data.offsetLeft!
    })

    const direction = computed<string>(() => {
      return isDirectY.value ? 'clientY' : 'clientX'
    })

    function moveReverse(size) {
      const {
        parentNode,
        offsetTop,
        offsetLeft,
        marginLeft,
        marginTop
      } = data

      const offset = props.top ?
        offsetTop! - marginTop :
        offsetLeft! - marginLeft

      parentNode!.style[reverseFrom.value] = `${ currentSize.value - size + offset }px`
    }

    function setOrEmitSize(size) {

      if (!props.emit) {
        data.parentNode!.style[styleProp.value] = size + 'px'
      } else {
        emit('size', size)
      }

      isNeedReverse.value && moveReverse(size)
    }

    function resize(e) {
      const startPoint = e[direction.value]

      let size

      if (isNeedReverse.value) {
        size = currentSize.value! - (startPoint - offset.value!)
      } else {
        size = currentSize.value! + (startPoint - currentSize.value! - offset.value!)
      }

      size > props.minSize ? setOrEmitSize(size) : false
    }

    function resetMinMaxStyles() {
      if (isDirectY.value) {
        data.parentNode!.style.maxHeight = ''
        data.parentNode!.style.minHeight = ''
      } else {
        data.parentNode!.style.maxWidth = ''
        data.parentNode!.style.minWidth = ''
      }
    }

    function computeParentNode() {
      const parent = resRef.value!.parentNode
      const {
        left,
        top,
        height,
        width,
        marginLeft,
        marginTop
      } = getComputedStyle(parent as HTMLElement)

      data.parentNode = parent as HTMLElement
      data.offsetTop = parseFloat(top)
      data.offsetLeft = parseFloat(left)
      data.marginLeft = parseFloat(marginLeft)
      data.marginTop = parseFloat(marginTop)
      data.parentHeight = parseFloat(height)
      data.parentWidth = parseFloat(width)
      data.offsetTop += data.marginTop
      data.offsetLeft += data.marginLeft
    }

    function disableSelection (e) {
      e.preventDefault()
    }

    function initResize(e) {
      if (!data.resized) {
        data.resized = true
        computeParentNode()
        resetMinMaxStyles()
      }
      resize(e)
    }

    function reset() {
      data.resized = false
      resetMinMaxStyles()
    }

    function onMouseup() {
      reset()
      removeHandlers()
    }

    function onMousedown() {
      document.addEventListener('mousemove', initResize)
      document.addEventListener('mouseup', onMouseup)
      document.addEventListener('selectstart', disableSelection)
    }

    function removeHandlers() {
      document.removeEventListener('mousemove', initResize)
      document.removeEventListener('mouseup', onMouseup)
      document.removeEventListener('selectstart', disableSelection)
    }

    onMounted(() => {
      computeParentNode()
    })

    return () => h('div', {
      class: {
        ...classes.value,
      },
      key: 'resize',
      ref: resRef,
      onMousedown,
    })
  },
})
