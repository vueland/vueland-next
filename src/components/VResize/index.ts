import VResize from './VResize'

export { VResize }
export default VResize
