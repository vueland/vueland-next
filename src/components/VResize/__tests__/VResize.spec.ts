import { mount, VueWrapper } from '@vue/test-utils'
import { VResize } from '../VResize'
import { VCard } from '../../VCard'
import { SetupContext } from 'vue'
import 'regenerator-runtime/runtime'

describe('VResize', () => {
  let mountInWrapFunction: (options?: any) => VueWrapper<any>
  let mountFunction: (options?: any) => VueWrapper<any>

  beforeEach(() => {
    mountInWrapFunction = (options = {}) => mount(VCard, {
      slots: {
        default: VResize.setup!(options, {} as SetupContext)
      },
      ...options,
    })
    mountFunction = (options = {}) => mount(VResize, {
      ...options
    })
  })

  it('should mount the component and match snapshot', () => {
    const cmp = mountFunction()

    expect(cmp.html()).toMatchSnapshot()
  })

  it('should set top prop and match snapshot', () => {
    const top = true
    const cmp = mountFunction({ propsData: { top } })

    expect(cmp.find('.v-resize').attributes().class).toContain('v-resize--top')
    expect(cmp.html()).toMatchSnapshot()
  })

  it('should set left prop and match snapshot', () => {
    const left = true
    const cmp = mountFunction({ propsData: { left } })

    expect(cmp.find('.v-resize').attributes().class).toContain('v-resize--left')
    expect(cmp.html()).toMatchSnapshot()
  })

  it('should set right prop and match snapshot', () => {
    const right = true
    const cmp = mountFunction({ propsData: { right } })

    expect(cmp.find('.v-resize').attributes().class).toContain('v-resize--right')
    expect(cmp.html()).toMatchSnapshot()
  })

  it('should set bottom prop and match snapshot', () => {
    const bottom = true
    const cmp = mountFunction({ propsData: { bottom } })

    expect(cmp.find('.v-resize').attributes().class).toContain('v-resize--bottom')
    expect(cmp.html()).toMatchSnapshot()
  })

  it('should set top prop and match snapshot', async () => {
    const cmp = mountInWrapFunction({
      top: true
    })
    const resizer = cmp.find('.v-resize')

    await resizer.trigger('mousedown')
    await resizer.trigger('mousemove')

    expect(cmp.html()).toMatchSnapshot()
  })
})
