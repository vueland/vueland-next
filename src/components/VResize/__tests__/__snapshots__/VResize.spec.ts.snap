// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VResize should mount the component and match snapshot 1`] = `
<div class="v-resize">
</div>
`;

exports[`VResize should set bottom prop and match snapshot 1`] = `
<div class="v-resize v-resize--bottom">
</div>
`;

exports[`VResize should set left prop and match snapshot 1`] = `
<div class="v-resize v-resize--left">
</div>
`;

exports[`VResize should set right prop and match snapshot 1`] = `
<div class="v-resize v-resize--right">
</div>
`;

exports[`VResize should set top prop and match snapshot 1`] = `
<div class="v-resize v-resize--top">
</div>
`;

exports[`VResize should set top prop and match snapshot 2`] = `
<div class="v-card"
     style="max-width: 350px;"
>
  <div class="v-resize v-resize--top">
  </div>
</div>
`;
